src/index/addrindex.cpp

```diff
@@ -0,0 +1,263 @@
+        for (size_t j = 0; j < tx.vout.size(); ++j) {
+            CScript script_pub_key = tx.vout[j].scriptPubKey;
+            DBKey key(DBKeyType::CREATED, GetAddrId(script_pub_key), COutPoint(tx_hash, j));
+            entries.emplace_back(key, std::make_pair(pos, script_pub_key));
+        }
+
+        // Skip coinbase inputs.
+        if (not_genesis_block && i > 0) {
+            const CTxUndo& tx_undo = block_undo.vtxundo[i-1];
+            for (size_t k = 0; k < tx.vin.size(); ++k) {
+                CScript spent_outputs_scriptpubkey = tx_undo.vprevout[k].out.scriptPubKey;
+                DBKey key(DBKeyType::SPENT, GetAddrId(spent_outputs_scriptpubkey), tx.vin[k].prevout);
+                entries.emplace_back(key, std::make_pair(pos, spent_outputs_scriptpubkey));
+            }
+        }
+        pos.nTxOffset += ::GetSerializeSize(tx, CLIENT_VERSION);
+    }
+
+    return m_db->WriteToIndex(entries);
+}
+
+bool AddrIndex::DB::WriteToIndex(const std::vector<std::pair<DBKey, DBValue>> &entries)
+{
+    CDBBatch batch(*this);
+    for (const auto& entry : entries) {
+        batch.Write(entry.first, entry.second);
+    }
+    return WriteBatch(batch);
+}
+
+// FindTxsByScript fills the spends_result vector with outpoints corresponding
+// to the output spent with the given script, and the transaction it was spent
+// in. creations_result is filled with outpoints for outputs created with this
+// script as their script pubkey, and the transactions they were created in.
+bool AddrIndex::FindTxsByScript(const CScript& script,
+                                std::vector<std::pair<COutPoint, std::pair<CTransactionRef, uint256>>> &spends_result,
+                                std::vector<std::pair<COutPoint, std::pair<CTransactionRef, uint256>>> &creations_result)
+{
+    auto db_entries = m_db->ReadAddrIndex(GetAddrId(script), script);
+    if (db_entries.size() == 0) return false;
+
+    for (const auto& entry : db_entries) {
+        DBKey key = entry.first;
+        CDiskTxPos pos = entry.second.first;
+
+        CAutoFile file(OpenBlockFile(pos, true), SER_DISK, CLIENT_VERSION);
+        if (file.IsNull()) {
+            return error("%s: OpenBlockFile failed", __func__);
+        }
+        CBlockHeader header;
+        CTransactionRef tx;
+        try {
+            file >> header;
+            if (fseek(file.Get(), pos.nTxOffset, SEEK_CUR)) {
+                return error("%s: fseek(...) failed", __func__);
+            }
+            file >> tx;
+        } catch (const std::exception& e) {
+            return error("%s: Deserialize or I/O error - %s", __func__, e.what());
+        }
+        std::pair<CTransactionRef, uint256> result =  std::make_pair(tx, header.GetHash());
+
+        // Place entry into correct vector depending on its type.
+        switch (key.m_key_type) {
+            case DBKeyType::SPENT:
+                spends_result.emplace_back(std::make_pair(key.m_outpoint, result));
```

Comment:
As pointed out by jnewbery in review club, I think this is wrong. 

The DB's key always represents the outpoint that *created* the scriptPubKey.  For spends_result, you want the txid of the transaction that *spends* the scriptPubKey.  So instead of the DB key's outpoint as the transaction id to be returned to the user, you want the transaction id of the transaction in the DB value.  Or, in the RPC code lines 314 and 315, use the txid from `spend.second`.